subsys/net/l2/ieee802154/ieee802154.c

```diff
@@ -92,9 +89,137 @@ static inline void ieee802154_acknowledge(struct net_if *iface, struct ieee80215
 
 	return;
 }
-#else
-#define ieee802154_acknowledge(...)
-#endif /* CONFIG_NET_L2_IEEE802154_ACK_REPLY */
+
+inline bool ieee802154_prepare_for_ack(struct net_if *iface, struct net_pkt *pkt,
+				       struct net_buf *frag)
+{
+	bool ack_required = ieee802154_is_ar_flag_set(frag);
+
+	if (ieee802154_get_hw_capabilities(iface) & IEEE802154_HW_TX_RX_ACK) {
+		return ack_required;
+	}
+
+	if (ack_required) {
+		struct ieee802154_fcf_seq *fs = (struct ieee802154_fcf_seq *)frag->data;
+		struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+		if (fs->sequence == 0) {
+			NET_ERR("Trying to send a packet requesting ACK but not containing a valid "
+				"sequence number.");
+			return false;
+		}
+
+		ctx->ack_seq = fs->sequence;
+		ctx->ack_received = false;
+		k_sem_init(&ctx->ack_lock, 0, K_SEM_MAX_LIMIT);
+
+		return true;
+	}
+
+	return false;
+}
+
+enum net_verdict ieee802154_handle_ack(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (ieee802154_get_hw_capabilities(iface) & IEEE802154_HW_TX_RX_ACK) {
+		__ASSERT_NO_MSG(ctx->ack_seq = 0U);
+		return NET_CONTINUE;
+	}
+
+	if (pkt->buffer->len == IEEE802154_ACK_PKT_LENGTH) {
+		uint8_t len = IEEE802154_ACK_PKT_LENGTH;
+		struct ieee802154_fcf_seq *fs;
+
+		fs = ieee802154_validate_fc_seq(net_pkt_data(pkt), NULL, &len);
+		if (!fs || fs->sequence != ctx->ack_seq) {
+			return NET_CONTINUE;
+		}
+
+		ctx->ack_received = true;
+		k_sem_give(&ctx->ack_lock);
+
+		/* TODO: Release packet in L2 as we're taking ownership. */
+		return NET_OK;
+	}
+
+	return NET_CONTINUE;
+}
+
+inline int ieee802154_wait_for_ack(struct net_if *iface, bool ack_required)
+{
+	struct ieee802154_context *ctx = net_if_l2_data(iface);
+
+	if (!ack_required || (ieee802154_get_hw_capabilities(iface) & IEEE802154_HW_TX_RX_ACK)) {
+		__ASSERT_NO_MSG(ctx->ack_seq = 0U);
+		return 0;
+	}
+
+	if (k_sem_take(&ctx->ack_lock, K_MSEC(10)) == 0) {
+		/* We reinit the semaphore in case ieee802154_handle_ack()
+		 * got called multiple times.
+		 */
+		k_sem_init(&ctx->ack_lock, 0, K_SEM_MAX_LIMIT);
+	}
+
+	ctx->ack_seq = 0U;
+
+	return ctx->ack_received ? 0 : -ETIME;
+}
+
+int ieee802154_radio_send(struct net_if *iface, struct net_pkt *pkt, struct net_buf *frag)
+{
+	bool ack_required;
+	int ret;
+
+	NET_DBG("frag %p", frag);
+
+	if (IS_ENABLED(CONFIG_NET_L2_IEEE802154_RADIO_CSMA_CA) &&
+	    ieee802154_get_hw_capabilities(iface) & IEEE802154_HW_CSMA) {
+		/* CSMA and retransmission in hardware */
```

Comment:
Could you elaborate more. In case of RF2xx it provides re-transmission when using this option currently. https://github.com/zephyrproject-rtos/zephyr/blob/135f5376c1e64b2f7ddef23f64f202d27f28674a/drivers/ieee802154/ieee802154_rf2xx.c#L658-L663

Reply:
Hmm, then it seems RF2xx and nRF5 drivers (I didn't anlyze others yet) interpret the capability differently - our driver only covers CSMA/CA for a single transmission. Perhaps we'd need another capability for retransmissions?

Reply:
@nandojve The standard distinguishes between retransmissions and CSMA/CA backoffs - they are totally unrelated in fact. In case of channel access failure, retransmission should not even be tried according to the standard. Retransmissions are only appropriate after CSMA/CA was successful, a transmission has taken place and an expected ACK did not arrive. Each re-transmission will require another round of CSMA/CA.

So IMO @rlubos is right: We need an "auto-retransmit" capability which signals to L2 that retransmissions are offloaded to the HW in case of RF2xx. This is not the case for all hardware. So these have to be separate capabilities.

BTW: Why is the driver not respecting user configuration re max retries, maxBE, minBE and CSMA/CA retries (aka maxBo - 1)? Setting them to fixed values looks like a bug to me.

Reply:
Fixed - I introduces an IEEE802154_HW_RETRANSMISSION capability and added it to all drivers that do retransmission.